logger/syslogish/server.go

```diff
@@ -0,0 +1,170 @@
+package syslogish
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/deis/deis/logger/drain"
+	"github.com/deis/deis/logger/storage"
+)
+
+var appRegex *regexp.Regexp
+
+func init() {
+	appRegex = regexp.MustCompile(`^.* ([-_a-z0-9]+)\[[a-z0-9-_\.]+\].*`)
+}
+
+// Server implements a UDP-based "syslog-like" server.  Like syslog, as described by RFC 3164, it
+// expects that each packet contains a single log message and that, conversely, log messages are
+// encapsulated in their entirety by a single packet, however, no attempt is made to parse the
+// messages received or validate that they conform to the specification.
+type Server struct {
+	conn           net.PacketConn
+	listening      bool
+	queue          chan string
+	storageAdapter storage.Adapter
+	drain          drain.LogDrain
+	adapterMutex   sync.RWMutex
+	drainMutex     sync.RWMutex
+}
+
+// NewServer returns a pointer to a new Server instance.
+func NewServer(bindHost string, bindPort int) (*Server, error) {
+	addr, err := net.ResolveUDPAddr("udp", fmt.Sprintf("%s:%d", bindHost, bindPort))
+	if err != nil {
+		return nil, err
+	}
+	c, err := net.ListenUDP("udp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &Server{conn: c, queue: make(chan string, 5)}, nil
+}
+
+// SetStorageAdapter permits a server's underlying storage.Adapter to be reconfigured (replaced)
+// at runtime.
+func (s *Server) SetStorageAdapter(storageAdapter storage.Adapter) {
+	// Get an exclusive lock before updating the internal pointer to the storage adapter.  Other
+	// goroutines holding read locks might depend on that pointer as it currently exists.
+	s.adapterMutex.Lock()
+	defer s.adapterMutex.Unlock()
+	s.storageAdapter = storageAdapter
+}
+
+// SetDrain permits a server's underlying drain.LogDrain to be reconfigured (replaced) at runtime.
+func (s *Server) SetDrain(drain drain.LogDrain) {
+	// Get an exclusive lock before updating the internal pointer to the log drain.  Other
+	// goroutines holding read locks might depend on that pointer as it currently exists.
+	s.drainMutex.Lock()
+	defer s.drainMutex.Unlock()
+	s.drain = drain
+}
+
+// Listen starts the server's main loop.
+func (s *Server) Listen() {
+	// Should only ever be called once
+	if !s.listening {
+		s.listening = true
+		go s.receive()
+		go s.process()
+		log.Println("syslogish server running")
+	}
+}
+
+func (s *Server) receive() {
+	// Make buffer the same size as the max for a UDP packet
+	buf := make([]byte, 65535)
+	for {
+		n, _, err := s.conn.ReadFrom(buf)
+		if err != nil {
+			log.Fatal("syslogish server read error", err)
+		}
+		message := strings.TrimSuffix(string(buf[:n]), "\n")
+		select {
+		case s.queue <- message:
+		default:
+		}
+	}
+}
+
+func (s *Server) process() {
+	for message := range s.queue {
```

Comment:
> It reads one message at a time off of the queue

Unavoidable, unfortunately.  To ensure multiple messages for a given app are stored in the same order they were sent (the same order in which the app produced them), messages have to be handled sequentially.

> and then locks at least three mutex objects

It's not actually that bad.  The locks acquired on adapters and drains are only read locks to ensure those components aren't nilled out by the configurer after we've checked that they aren't already nil.  Read locks aren't deleterious to performance in the way a write lock is since you rarely block waiting for a read lock.  As for the mutexes used within the storage adapters, we always check whether we need a lock before we bother to get one.  The conditions where one is needed are the first message for a new app or the destruction of logs for an app that's also been destroyed.  Both of these are relatively rare events.

> and the buffer will fill quickly. Then your sending process will begin to block...

The write to the internal channel is non-blocking.  If the buffer that is that channel fills up, messages are dropped right there so the UDP listener's own queue doesn't fill up.

> increasing channel buffer size

Bumped to 1000 per your recommendation.

> writing to an intermediate in-memory cache

That's the point of the internal channel / queue, so this is already being done.